src/hotspot/share/classfile/protectionDomainCache.cpp

```diff
@@ -58,29 +60,67 @@ void ProtectionDomainCacheTable::trigger_cleanup() {
 }
 
 class CleanProtectionDomainEntries : public CLDClosure {
+  GrowableArray<ProtectionDomainEntry*>* _delete_list;
+ public:
+  CleanProtectionDomainEntries(GrowableArray<ProtectionDomainEntry*>* delete_list) :
+                               _delete_list(delete_list) {}
+
   void do_cld(ClassLoaderData* data) {
     Dictionary* dictionary = data->dictionary();
     if (dictionary != NULL) {
-      dictionary->clean_cached_protection_domains();
+      dictionary->clean_cached_protection_domains(_delete_list);
     }
   }
 };
 
+class HandshakeForPD : public HandshakeClosure {
+ public:
+  HandshakeForPD() : HandshakeClosure("HandshakeForPD") {}
+
+  void do_thread(Thread* thread) {
+    log_trace(protectiondomain)("HandshakeForPD::do_thread: thread="
+                                INTPTR_FORMAT, p2i(thread));
+  }
+};
+
+static void purge_deleted_entries(GrowableArray<ProtectionDomainEntry*>* delete_list) {
+  // If there are any deleted entries, Handshake-all then they'll be
+  // safe to remove since traversing the pd_set list does not stop for
+  // safepoints and only JavaThreads will read the pd_set.
```

Comment:
I do not understand what you are trying to do here. This is basically a no-op handshake with every thread? Does all the threads remain in the handshake until the last one is processed or do they proceed through one at a time? The former seems better suited for a safepoint operation. The latter means this doesn't work as they may have moved on to a new removal.

Reply:
Only the ServiceThread calls this code, so they won't be removed again.  This is modeled after the ObjectMonitor code that does the same thing.  All the threads have to hit a handshake at a safepoint polling place, which they cannot do while reading the list.  After all the threads have hit the handshake, no thread can see the old entries on the list.
I thought that's what I said in the comment.  What should I say to make this clearer?

Reply:
See src/hotspot/share/runtime/synchronizer.cpp:

```
size_t ObjectSynchronizer::deflate_idle_monitors() {
<snip>
    if (current->is_Java_thread()) {
<snip>
      // A JavaThread needs to handshake in order to safely free the
      // ObjectMonitors that were deflated in this cycle.
      HandshakeForDeflation hfd_hc;
      Handshake::execute(&hfd_hc);
```